### ISSUE TITLE
Realm initial sync stucking on 'Pending' when a collection is big

### ISSUE THREAD

**Issue by aureliopetrone** [realm/realm-js#2517]
I have a database of ~250mb. When I open a synced realm on the client I subscribe to a class of objects and I add a listener to wait for the sync to complete.

My problem is that, on Android simulator, the subscription doesn't go on and the last state I can track is 'Pending'.

This issue seems to appears only on Android Emulator but I had the same issue, with a ~1Gb database also on a Galaxy S9. **I'm worried that this isn't the best way to do initial sync but I can't find any solutions provided by Realm to such a scenario.** 

## Code Sample

```

await this.susbscribeAndSyncTo(this.realm_instance.objects("Cliente"), "Cliente") // This is where the sync begin

static susbscribeAndSyncTo = async (object, object_name) => {

        var subscription = object.subscribe()

        return new Promise((resolve, reject) => {
            subscription.addListener((subscription, state) => {
                if (this.checkSubscriptionState(state, object_name)) {
                    try {
                        subscription.removeAllListeners()
                    } catch (e) {
                        console.log(e.message)
                    }
                    resolve(true);
                }
            })
        });

    }
    static checkSubscriptionState = (state, object_type) => {

        switch (state) {
            case Realm.Sync.SubscriptionState.Creating:
                logIfDebug(this.DEBUG,"Creo la subscription per", object_type)
                // The subscription has not yet been written to the Realm
                break;
            case Realm.Sync.SubscriptionState.Pending:
                logIfDebug(this.DEBUG,"Attendo l'arrivo dei dati", object_type)
                // The subscription has been written to the Realm and is waiting
                // to be processed by the server
                break;
            case Realm.Sync.SubscriptionState.Complete:
                logIfDebug(this.DEBUG,"I dati sono sincronizzati con il server", object_type)
                // The subscription has been processed by the server and all objects
                // matching the query are in the local Realm
                return true

                break;
            case Realm.Sync.SubscriptionState.Invalidated:
                logIfDebug(this.DEBUG,"La sottoscrizione ai dati è stata invalidata", object_type)
                // The subscription has been removed
                break;
            case Realm.Sync.SubscriptionState.Error:
                logIfDebug(this.DEBUG,"C'è stato un errore nella sincronizzazione con il server", object_type)
                break;

            default:
                logIfDebug(this.DEBUG,"Mmmm")
                break;
        }

        return false

    }
```

## Version of Realm and Tooling

- Realm JS SDK Version: 2.29.2
- Node or React Native: 0.60
- Client OS & Version: Android Emulator
- Which debugger for React Native: ?/None

**Comment by bmunkholm** [comment]
Could I ask what data you are syncing that is so big? Not big blobs right?

**Comment by aureliopetrone** [comment]
Sure, the database is made of 4-5 collections. The sync starts with a collection of ~6000 objects but these objects have others object as children that are synced...So it's like

Ob. Class A ~6000 > Ob. Class B ~ 1000 > Ob. Class C~ 5000 > Ob. Class D ~ 25000

This is a schema exeample:

```

[...]
export const Cliente = {
    name: "Cliente",
    primaryKey: "id",
    properties: {
      id: "string",
      tipologia: "string?",
      owner_ob: "Agente",
      brand: 'string',
      agenzia: "Agenzia[]",
      ragionesociale: "string?",
      piva: "string?",
      nazione: "string?",
      indirizzo: "string?",
      citta: "string?",
      provincia: "string?",
      cap: "string?",
      email: "string?",
      email2: "string?",
      email3: "string?",
      telefono: "string?",
      telefono2: "string?",
      telefono3: "string?",
      telefono3: "string?",
      timestamp: "date?",
      fax: "string?",
      web: "string?",
      referente: "string?",
      status: "string?",
      codicecliente: "string?",
      pictures: "Pic[]",
      schede: "Scheda[]",
      books: "Book[]",
      permissions: '__Permission[]'
    }
  };
  
  export const Scheda = {
    name: "Scheda",
    primaryKey: "id",
    properties: {
      id: "string",
      owner_ob: "Agente",
      cliente: {type: 'linkingObjects', objectType: 'Cliente', property: 'schede'},
      preferenze : 'PreferenzaProdotto[]',
      timestamp: "date",
      tipologia: "string",
      note: 'string',
      permissions: '__Permission[]'
    }
  };
  
  export const Agente = {
    name: "Agente",
    primaryKey: "id",
    properties: {
      id: "string",
      user: "__User",
      nome: "string",
      agenzia: {type: 'linkingObjects', objectType: 'Agenzia', property: 'agenti'},
      books: {type: 'linkingObjects', objectType: 'Book', property: 'owner_ob'},
      schede: {type: 'linkingObjects', objectType: 'Scheda', property: 'owner_ob'},
      clienti: {type: 'linkingObjects', objectType: 'Cliente', property: 'owner_ob'},
      progressivo: 'int',
      sigla: 'string',
      imgprofilo: "Pic",
      miniatura: "Pic",
      ruolo: 'string',
      telefono: 'string',
      email: 'string'
    }
  };
  
  export const Agenzia = {
    name: "Agenzia",
    primaryKey: "id",
    properties: {
      id: "string",
      nome: "string",
      clienti: {type: 'linkingObjects', objectType: 'Cliente', property: 'agenzia'},
      agenti: "Agente[]",
    }
  };
  
  
  export const PreferenzaProdotto = {
    name: "PreferenzaProdotto",
    primaryKey: "id",
    properties: {
      id: "string",
      timestamp: "date",
      prodotto: "Prodotto?",
      variantipreferenzaprodotto: "VariantePreferenzaProdotto[]",
      prezzo_valore: "string?",
      prezzo_unita: "string?",
      prezzo_standard: "bool?",
      permissions: '__Permission[]',
      scheda: {type: 'linkingObjects', objectType: 'Scheda', property: 'preferenze'},
      note: 'string?'
    }
  };
  
  export const VariantePreferenzaProdotto = {
    name: "VariantePreferenzaProdotto",
    primaryKey: "id",
    properties: {
      id: "string",
      colore: "string",
      quantita_valore: "string?",
      quantita_unita: "string?",    
      timestamp: "date",
      permissions: '__Permission[]',
      preferenzaprodotto: {type: 'linkingObjects', objectType: 'PreferenzaProdotto', property: 'variantipreferenzaprodotto'},
    }
  };
  
  export const Prodotto = {
    name: "Prodotto",
    primaryKey: "id",
    properties: {
      id: "string",
      colori: "string[]",   
      nome: "string",
      tipologia: "string",
      prezzo1: "string",
      collezione: 'string',
      prezzo2: "string",
      picture: "Pic",
      pictures: "Pic[]",
      cartellariferimento: 'string',
      tipocartella: 'string?',
      scansionicartella: 'data?[]'
    }
  };

[...]
```

I start the sync subscribing to 'Cliente'.

Thanks @bmunkholm

**Comment by bmunkholm** [comment]
Thanks @aureliopetrone. Don't know if you noticed in the template we make it clear that this repo is only for bugs or feature requests. So this question belongs better in our Forum.
That said, a few questions to consider:
- Don't store pictures in (any) database. Store them in a filesystem and retrieve them as needed. Only store the meta data.
- What happens if you only have very little data? Does everything work as expected? Is it only "stuck" because the initial sync takes too long?
- If data isn't changing too much, could you bundle the initial data with the app?
- Do you need to provide all data to all phones? Or is a subset enough initially?

But again, unless you can show there is a bug (with a small reproducible testcase), please move this discussion to the Forum with some of the questions answered. 

Thanks!

**Comment by aureliopetrone** [comment]
Hi @bmunkholm,

If I have little data it works. It works even if I change

await this.susbscribeAndSyncTo(this.realm_instance.objects("Cliente"), "Cliente")
      
to

await this.susbscribeAndSyncTo(this.realm_instance.objects("Cliente").filtered("nazione == 'ITA'"), "Cliente")
        await this.susbscribeAndSyncTo(this.realm_instance.objects("Cliente").filtered("nazione != 'ITA'"), "Cliente")

so I think it's a bug that shows when collections are too big...

**Comment by bmunkholm** [comment]
What makes you think it's a bug and not just because it takes a long time?

**Comment by aureliopetrone** [comment]
Because I waited for about 2 hrs with the App open :)

On other devices (real ones) it takes 30s to sync

**Comment by bmunkholm** [comment]
So are you are ok with the performance and functionality on real devices. It's just in the Emulator you notice a problem?

**Comment by aureliopetrone** [comment]
In this case yes. But as I mentioned I noticed this problem also on a real device (Samsung Galaxy S9+) with a bigger database I tested to understand how big a realm could be.  I fixed by dividing the subscription in smaller ones (as I showed in my previous post) but I don't think that it's a good solution but more a 'workaround'.

I've seen that in the 2.29.0 has been introduced a new feature fr waiting for the database to download initial data before opening ( Realm.Sync.SyncConfiguration.newRealmFileBehavior ) but:

1) How a big subscription like this would work with that
2) There is no documentation at all, the link to docs in Changelog (on Github) is broken and API reference for realm js doesn't mention it

It's not a big problem in my case but I think that for the ones using very big databases it could be.

**Comment by bmunkholm** [comment]
It would be very helpful if you could make a minimal repro case that can reproduce this and show the difference in time depending on the size.

1. There are no size limitations, except what the device can hold. But if you want to sync that much, shouldn't you use full sync? What are you not syncing?
2. Oh, happy you noticed that! That should get fixed.
@kneth @cbush 
See [Changelog](https://github.com/realm/realm-js/blob/master/CHANGELOG.md#2290-release-notes-2019-7-3) where link is wrong.

**Comment by aureliopetrone** [comment]
I don't use full sync because I find the fine-grained permission system convenient for my use-case.
I surely can but since I have near deadlines I will do that in october ;)

**Comment by kneth** [comment]
I have added a link in the changelog (see https://github.com/realm/realm-js/pull/2523) but we need to update the documentation. Since all SDKs have this functionality, I suggest it should be documented in https://docs.realm.io/sync/using-synced-realms/setting-up-your-realms.

**Comment by cbush** [comment]
It seems not all SDKs have this functionality yet. I can update the JS code examples for now (see [draft](https://app.gitbook.com/@realm/s/sync/~/drafts/-Lpii_YJJefPFL6JDSEj/primary/using-synced-realms/setting-up-your-realms#asynchronously-opening-a-realm)), but updating the general docs will just make it more confusing. I will hold off for now.

**Comment by kneth** [comment]
@cbush 
It has been released in Realm Cocoa v3.16.0 and Realm Java v5.13.0.
